src/test/java/com/puppycrawl/tools/checkstyle/JavaAstVisitorTest.java

```diff
@@ -0,0 +1,182 @@
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import static com.google.common.truth.Truth.assertWithMessage;
+
+import java.io.File;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.grammar.java.CheckstyleJavaParserBaseVisitor;
+
+public class JavaAstVisitorTest {
+
+    /**
+     * If a visit method is not overridden, we should explain why we do not 'visit' the
+     * parsetree at this node and construct an AST. Reasons could include that we have
+     * no terminal symbols (tokens) in the corresponding production rule, or that
+     * we handle the construction of this particular AST in it's parent node. If we
+     * have a production rule where we have terminal symbols (tokens), but we do not build
+     * an AST from tokens in the rule context, the rule is extraneous.
+     */
+    private static final List<String> VISIT_METHODS_NOT_OVERRIDDEN = Arrays.asList(
+            // no tokens in production rule, so no AST to build
+            "visitClassOrInterfaceOrPrimitiveType",
+            "visitNonWildcardTypeArgs",
+            "visitStat",
+            "visitAnnotationConstantRest",
+            "visitSwitchLabeledRule",
+            "visitLocalVariableDeclaration",
+            "visitTypes",
+            "visitSwitchStat",
+            "visitSwitchPrimary",
+            "visitClassDef",
+            "visitInterfaceMemberDeclaration",
+            "visitMemberDeclaration",
+            "visitLiteralPrimary",
+            "visitPatternDefinition",
+            "visitLocalType",
+            "visitLocalTypeDeclaration",
+            "visitRecordBodyDeclaration",
+            "visitResource",
+            "visitVariableInitializer",
+            "visitLambdaBody",
+
+            // AST built in parent rule
+            "visitCreatedNameExtended",
+            "visitSuperSuffixSimple",
+            "visitFieldAccessNoIdent",
+            "visitClassType",
+            "visitClassOrInterfaceTypeExtended"
+    );
+
+    @Test
+    public void testAllVisitMethodsAreOverridden() {
```

Comment:
With ANTLR2, we relied on the automatic AST generation from file parsing; with ANTLR4 we have to generate our own. This leaves room for error; we could add a production rule to the grammar, but not visit the rule context and create an AST. I have created this test to make sure that this doesn't happen; anytime a new rule or labeled alternative is created, a new `visit...` method is generated in `CheckstyleJavaParserBaseVisitor` automatically. This test checks that all `visit` methods are overridden, or that the method name is explicitly added to `VISIT_METHODS_NOT_OVERRIDDEN` list with a comment to explain why we don't override the visit method and build an AST.